subsys/net/lib/sockets/socketpair.c

```diff
@@ -0,0 +1,1087 @@
+	if (spair->remote == -1) {
+		errno = ENFILE;
+		goto cleanup;
+	}
+
+	z_finalize_fd(spair->remote, spair,
+		      (const struct fd_op_vtable *)&spair_fd_op_vtable);
+
+	NET_DBG("%p fd: %d sem: %p vtable: %p recv_q: %p buf: %p "
+		"write: %p read: %p", spair, spair->remote, &spair->sem,
+		&spair_fd_op_vtable, &spair->recv_q, spair->buf,
+		&spair->write_signal, &spair->read_signal);
+
+	goto out;
+
+cleanup:
+	spair_delete(spair);
+	spair = NULL;
+
+out:
+	return spair;
+}
+
+int z_impl_zsock_socketpair(int family, int type, int proto, int *sv)
+{
+	int res;
+	size_t i;
+	struct spair *obj[2] = {};
+
+	if (family != AF_UNIX) {
+		errno = EAFNOSUPPORT;
+		res = -1;
+		goto out;
+	}
+
+	if (!(type == SOCK_STREAM || type == SOCK_RAW)) {
+		errno = EPROTOTYPE;
+		res = -1;
+		goto out;
+	}
+
+	if (proto != 0) {
+		errno = EPROTONOSUPPORT;
+		res = -1;
+		goto out;
+	}
+
+	if (!sv) {
+		/* not listed in the normative standard, but probably safe */
+		errno = EINVAL;
+		res = -1;
+		goto out;
+	}
+
+	for (i = 0; i < 2; ++i) {
+		obj[i] = spair_new();
+		if (!obj[i]) {
+			res = -1;
+			goto cleanup;
+		}
+	}
+
+	for (i = 0; i < 2; ++i) {
+		sv[i] = obj[i]->remote;
+	}
+
+	/* connect the two endpoints */
+	swap32(&obj[0]->remote, &obj[1]->remote);
+
+	k_sem_give(&obj[0]->sem);
+	k_sem_give(&obj[1]->sem);
+
+	res = 0;
+
+	goto out;
+
+cleanup:
+	for (i = 0; i < 2; ++i) {
+		spair_delete(obj[i]);
+	}
+
+out:
+	return res;
+}
+
+#ifdef CONFIG_USERSPACE
+static inline int z_vrfy_zsock_socketpair(int family, int type, int proto,
+					  int *sv)
+{
+	int ret;
+	int tmp[2];
+
+	ret = z_impl_zsock_socketpair(family, type, proto, tmp);
+	Z_OOPS(z_user_to_copy(sv, tmp, sizeof(tmp));
+	return ret;
+}
+
+#include <syscalls/zsock_spair_mrsh.c>
+#endif /* CONFIG_USERSPACE */
+
+/**
+ * Write data to one end of a @ref spair
+ *
+ * Data written on one file descriptor of a socketpair can be read at the
+ * other end using common POSIX calls such as read(2) or recv(2).
+ *
+ * If the underlying file descriptor has the @ref O_NONBLOCK flag set then
+ * this function will return immediately. If no data was written on a
+ * non-blocking file descriptor, then -1 will be returned and @ref errno will
+ * be set to @ref EAGAIN.
+ *
+ * Blocking write operations occur when the @ref O_NONBLOCK flag is @em not
+ * set and there is insufficient space in the @em remote @ref spair.pipe.
+ *
+ * Such a blocking write will suspend execution of the current thread until
+ * one of two possible results is received on the @em remote
+ * @ref spair.read_signal:
+ *
+ * 1) @ref SPAIR_SIG_DATA - data has been read from the @em remote
+ *    @ref spair.pipe. Thus, allowing more data to be written.
+ *
+ * 2) @ref SPAIR_SIG_CANCEL - the @em remote socketpair endpoint was closed
+ *    Receipt of this result is analagous to SIGPIPE from POSIX
+ *    ("Write on a pipe with no one to read it."). In this case, the function
+ *    will return -1 and set @ref errno to @ref EINTR.
```

Comment:
Oh no, "EINTR" is only "interrupted system call" [by signal]. As we don't have POSIX signals in Zephyr, we never return EINTR from POSIX funcs. Please do some research what the right error should be, most likely EPIPE indeed.